File: packages/runtime-native/tests/pump_silence_test.cpp
// Pump-silence observer contract: no display, no GPU, no runtime.
//
// Proves the injected-clock logic of `mystral/pump_silence.h` directly:
// inter-entry retention with no filter and no windowing, the 64-cap with
// overflow counting, the once-only endpoint, and the fail-closed empty flush.
// Real-host retention (multi-pump gap, shutdown flush, SIGKILL absence) lives
// in `tests/pump-silence.test.mjs`, which drives the built desktop binary.

#include "mystral/pump_silence.h"

#include <cmath>
#include <cstdio>
#include <string>

namespace {

int failures = 0;

void expect(bool condition, const std::string& what, const std::string& detail = "") {
    if (condition) {
        std::printf("ok - %s\n", what.c_str());
        return;
    }
    failures += 1;
    std::printf("FAIL - %s%s%s\n", what.c_str(), detail.empty() ? "" : ": ", detail.c_str());
}

}  // namespace

int main() {
    // Inter-entry gap: 0 -> 10 -> 410 retains the 400 ms stall, keeps the max.
    {
        mystral::PumpSilenceObserver observer;
        observer.notePumpEntry(0.0);
        observer.notePumpEntry(10.0);
        observer.notePumpEntry(410.0);
        expect(observer.pumpCount() == 3, "three entries counted");
        expect(observer.maxGapMs() == 400.0, "max gap is the 400 ms stall");
        expect(observer.firstPumpAtMs() == 0.0, "first pump anchors process-to-first-pump");
        expect(observer.retainedLongGaps() == 1, "one long gap retained");
    }
    // Never-entered control: the empty observer stays empty, so a missing
    // observation can never read as a small-but-present maximum.
    {
        mystral::PumpSilenceObserver empty;
        expect(empty.pumpCount() == 0, "never-entered pump stays at zero");
        expect(empty.maxGapMs() == 0.0, "never-entered max stays zero");
        expect(empty.retainedLongGaps() == 0, "never-entered retains nothing");
    }
    // Cap: the global max survives past 64 retained gaps; overflow is counted.
    {
        mystral::PumpSilenceObserver observer;
        observer.notePumpEntry(0.0);
        double at = 0.0;
        for (int i = 0; i < 70; ++i) {
            at += 300.0;
            observer.notePumpEntry(at);
        }
        expect(observer.maxGapMs() == 300.0, "max survives past the retain cap");
        expect(observer.retainedLongGaps() == mystral::PumpSilenceObserver::kMaxRetainedGaps,
               "retained gaps stop at the cap");
        expect(observer.droppedLongGaps() == 6, "overflow counted, not dropped silently",
               std::to_string(observer.droppedLongGaps()));
    }
    // Endpoint: once-only. A first-frame flush suppresses the shutdown one,
    // so a launch emits exactly one line; shutdown still flushes a run that
    // never presented.
    {
        mystral::PumpSilenceObserver observer;
        observer.notePumpEntry(100.0);
        observer.flush(150.0);
        expect(observer.flushed(), "first flush marks the endpoint");
        observer.flush(999.0);
        expect(observer.flushed(), "second flush is a no-op");
        mystral::PumpSilenceObserver unpresented;
        unpresented.notePumpEntry(50.0);
        expect(!unpresented.flushed(), "shutdown path still open without a present");
        unpresented.flush(400.0);
        expect(unpresented.flushed(), "shutdown flush closes the endpoint");
    }
    // Non-positive gaps are not intervals: a backwards clock step advances
    // the counters but must not move the maximum or the retained list.
    {
        mystral::PumpSilenceObserver observer;
        observer.notePumpEntry(100.0);
        observer.notePumpEntry(90.0);
        observer.notePumpEntry(500.0);
        expect(observer.pumpCount() == 3, "backwards step still counts entries");
        expect(observer.maxGapMs() == 410.0, "backwards step does not set the max",
               std::to_string(observer.maxGapMs()));
        expect(observer.retainedLongGaps() == 1, "backwards step is not retained");
    }
    // Requested snapshot: non-consuming. A displacement-correlated endpoint
    // must not suppress the once-only diagnostic lines or reset counters.
    {
        mystral::PumpSilenceObserver observer;
        observer.notePumpEntry(10.0);
        observer.notePumpEntry(20.0);
        const std::string first = observer.snapshot(30.0);
        const std::string second = observer.snapshot(40.0);
        expect(!observer.flushed(), "snapshots never mark the endpoint");
        expect(observer.pumpCount() == 2, "snapshots preserve counter state");
        expect(first.find("\"atMs\":30.000") != std::string::npos, "snapshot stamps the request time",
               first);
        expect(second.find("\"atMs\":40.000") != std::string::npos, "later snapshot moves forward",
               second);
        expect(first.find("\"trailingGapMs\":10.000") != std::string::npos,
               "snapshot measures the trailing interval", first);
    }
    if (failures == 0) std::printf("pump-silence contract passed\n");
    return failures == 0 ? 0 : 1;
}

File: packages/runtime-native/tests/pump-silence.test.mjs
// PRD-360 pump-silence observer: real-host green proof.
//
// The observer (`include/mystral/pump_silence.h`) stamps every pollEvents()
// entry on the coldStartNowMs() clock and emits one TN_PUMP_SILENCE line at a
// bounded endpoint (first present, else shutdown). HostGapMeter is rAF-keyed,
// drops >2s periods and gates on a full window: unsuitable, and untouched.
// Missing observation (no line, or observed:false) is failure, never a pass.

import assert from "node:assert/strict";
import { execFileSync, spawn } from "node:child_process";
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { test } from "vitest";
import { makeTempDirSync } from "../../../test-support/temp-dir.js";

const ROOT = join(import.meta.dirname, "..");
const RUNTIME_SRC = readFileSync(join(ROOT, "src", "runtime.cpp"), "utf8");
const PRES_SRC = readFileSync(join(ROOT, "src", "webgpu", "bindings_presentation.cpp"), "utf8");
const HEADER = readFileSync(join(ROOT, "include", "mystral", "pump_silence.h"), "utf8");

function findBinary() {
  for (const candidate of [
    join(ROOT, "build", "tn-linux", "mystral"),
    join(ROOT, "build", "tn-linux-quickjs", "mystral"),
  ]) {
    try {
      readFileSync(candidate);
      return candidate;
    } catch {
      /* try next */
    }
  }
  throw new Error("TN_PUMP_PROBE_NO_BINARY: build the desktop host first (pnpm native:build)");
}

// Runs the real desktop host. Never xvfb-run (its cleanup kill replaces the
// command status); SDL_VIDEODRIVER=x11 is what verify-desktop-core.mjs sets.
// `screenshot`: the presenting path, so the bounded first-frame endpoint fires.
// `plain`: `run` without `--screenshot`, so the main-loop-exit endpoint fires
// (a never-presenting run proves the shutdown flush).
function runHost(bundlePath, mode = "screenshot") {
  const dir = makeTempDirSync("tn-pump-");
  const shot = join(dir, "shot.png");
  const xvfb = join(ROOT, "..", "..", "scripts", "xvfb.sh");
  const args =
    mode === "screenshot"
      ? ["run", bundlePath, "--screenshot", shot, "--frames", "5"]
      : ["run", bundlePath];
  try {
    return execFileSync("sh", [xvfb, "env", "SDL_VIDEODRIVER=x11", findBinary(), ...args], {
      cwd: dir,
      encoding: "utf8",
      timeout: 120_000,
    });
  } catch (error) {
    return `${error.stdout ?? ""}\n${error.stderr ?? ""}`;
  }
}

function runHostKill(bundleJs) {
  // SIGKILL mid-run: no graceful exit is possible, so no endpoint line can
  // exist. The harness must treat the missing line as failure, never a pass.
  const dir = makeTempDirSync("tn-pump-kill-");
  const bundle = join(dir, "game.js");
  const hostPidFile = join(dir, "host.pid");
  writeFileSync(bundle, bundleJs);
  const xvfb = join(ROOT, "..", "..", "scripts", "xvfb.sh");
  return new Promise((resolve) => {
    const chunks = [];
    const controller = 'printf "%s\\n" "$$" > "$1"; shift; exec "$@"';
    const child = spawn(
      xvfb,
      [
        "env",
        "SDL_VIDEODRIVER=x11",
        "sh",
        "-c",
        controller,
        "tn-pump-kill",
        hostPidFile,
        findBinary(),
        "run",
        bundle,
      ],
      { cwd: dir },
    );
    child.stdout.on("data", (chunk) => chunks.push(chunk));
    child.stderr.on("data", (chunk) => chunks.push(chunk));

    const killHost = () => {
      try {
        const hostPid = Number(readFileSync(hostPidFile, "utf8"));
        if (!Number.isInteger(hostPid) || hostPid <= 0) throw new Error("invalid host pid");
        process.kill(hostPid, "SIGKILL");
        return true;
      } catch {
        return false;
      }
    };

    setTimeout(() => {
      if (!killHost()) child.kill("SIGTERM");
    }, 3000);
    setTimeout(() => {
      killHost();
      resolve(Buffer.concat(chunks).toString("utf8"));
    }, 6000);
  });
}

const SMOKE = join(ROOT, "..", "..", "examples", "native-smoke", "dist", "native-smoke.js");

function pumpLine(log) {
  const line = log.split("\n").find((entry) => entry.includes("TN_PUMP_SILENCE:"));
  assert.ok(line, "TN_PUMP_SILENCE_UNOBSERVED: missing observation must fail, never pass");
  return JSON.parse(line.slice(line.indexOf("{")));
}

function endpointSnapshots(log) {
  return log
    .split("\n")
    .filter((entry) => entry.includes("TN_PUMP_ENDPOINT:"))
    .map((line) => JSON.parse(line.slice(line.indexOf("{"))));
}

// Drives the real host through an actual mailbox respond() with a
// displacement-shaped payload, like device.ts dispatch() after sampling.
// Returns the host log plus the response path for coordinator correlation.
function runMailboxRespond({ stallBeforeRespondMs = 0 } = {}) {
  const dir = makeTempDirSync("tn-pump-ep-");
  const root = join(dir, "mbox");
  mkdirSync(root, { recursive: true });
  const res = join(root, "response.json");
  const bundle = join(dir, "game.js");
  const spin =
    stallBeforeRespondMs > 0
      ? `const spinEnd = Date.now() + ${stallBeforeRespondMs}; while (Date.now() < spinEnd) {}`
      : "";
  writeFileSync(
    bundle,
    [
      `globalThis.TN_PLAYTEST_MAILBOX=${JSON.stringify({ request: join(root, "request.json"), response: res })};`,
      "let n = 0;",
      "const id = setInterval(() => {",
      "  n += 1;",
      "  if (n === 3) {",
      `    ${spin}`,
      `    __THREENATIVE_NATIVE__.playtest.respond(${JSON.stringify(res)}, JSON.stringify({ id: "sample-1", result: { entities: { player: { position: [0.41, -0.02, -0.46] } } } }));`,
      "  }",
      "  if (n >= 5) { clearInterval(id); process.exit(0); }",
      "}, 50);",
    ].join("\n"),
  );
  return { log: runHost(bundle, "plain"), res };
}

test("entry stamp sits at the pollEvents() head, ahead of every early return", () => {
  const pollStart = RUNTIME_SRC.indexOf("bool pollEvents() override");
  assert.ok(pollStart >= 0, "missing pollEvents");
  const entry = RUNTIME_SRC.slice(pollStart, pollStart + 600);
  assert.match(entry, /pumpSilence\(\)\.notePumpEntry\(coldStartNowMs\(\)\)/u);
  assert.ok(
    entry.indexOf("notePumpEntry") < entry.indexOf("hostGapMeter_.begin"),
    "the pump stamp must precede the kEvents bracket",
  );
});

test("the observer keeps every long gap: no hitch filter, no windowing", () => {
  assert.doesNotMatch(HEADER, /kHitchPeriodMicros|kWindow|dropPartialFrame|noteRafBegin/u);
  assert.match(HEADER, /kLongGapMs = 250\.0/u);
  assert.match(HEADER, /longGaps_\.push_back\(gapMs\)/u);
});

// Injected-clock unit logic moved to the registered CTest contract
// (`tests/pump_silence_test.cpp`, target `threenative-pump-silence-test`):
// inter-entry retention, cap overflow, once-only endpoint, empty control.
// This file keeps the real-host probes only.

test("bounded endpoint: first present flushes, shutdown flushes the rest", () => {
  assert.match(
    PRES_SRC,
    /pumpSilence\(\)\.flush\(mystral::coldStartNowMs\(\)\)/u,
    "first-frame endpoint must flush beside first_frame",
  );
  const shutdown = RUNTIME_SRC.slice(
    RUNTIME_SRC.indexOf("void shutdown()"),
    RUNTIME_SRC.indexOf("void shutdown()") + 500,
  );
  assert.match(shutdown, /pumpSilence\(\)\.flush\(coldStartNowMs\(\)\)/u);
  assert.match(HEADER, /if \(flushed_\) return;/u, "exactly one line per launch");
});

test("respond() stamps a non-consuming displacement-correlated endpoint", () => {
  assert.match(
    RUNTIME_SRC,
    /pumpSilence\(\)\.snapshot\(coldStartNowMs\(\)\)/u,
    "respond() must snapshot on the launch clock",
  );
  assert.match(
    RUNTIME_SRC,
    /TN_PUMP_ENDPOINT/u,
    "endpoint rides the existing diagnostic transport",
  );
});

test("real host: the pump is observed at the bounded endpoint", () => {
  // The smoke bundle presents, so the first-frame endpoint fires. A rAF-keyed
  // meter would need frame callbacks; this observer needs only pump entries.
  const payload = pumpLine(runHost(SMOKE));
  assert.equal(payload.observed, true);
  assert.ok(payload.pumpCount >= 1, "at least one pump entry before the endpoint");
  assert.ok(payload.firstPumpAtMs >= 0, "launch-relative first-pump stamp");
  assert.ok(payload.trailingGapMs >= 0, "trailing interval measured, not dropped");
});

test("real host: a deliberate 400ms stall is retained, not filtered", () => {
  // The spin runs INSIDE a pump iteration (timers segment). Screenshot mode
  // presents on pump 1, so it lands in the trailing interval; the plain-mode
  // timer run below proves the inter-entry half.
  const dir = makeTempDirSync("tn-pump-stall-");
  const stalled = join(dir, "stall.js");
  const smokeSrc = readFileSync(SMOKE, "utf8");
  writeFileSync(
    stalled,
    [
      "setTimeout(() => { const end = Date.now() + 400; while (Date.now() < end) {} }, 10);",
      smokeSrc,
    ].join("\n"),
  );
  const payload = pumpLine(runHost(stalled));
  assert.ok(
    payload.maxGapMs >= 250 ||
      payload.trailingGapMs >= 250 ||
      payload.longGaps.some((gap) => gap >= 250),
    `stall discarded: ${JSON.stringify(payload)}`,
  );
});

test("real host: multi-pump inter-entry >250ms gap is retained", () => {
  // Plain `run` (no screenshot gate): a 50 ms timer keeps pumps coming, a
  // 400 ms spin at tick 6 forces a genuine inter-entry gap. Proves retention
  // independently of rAF/presentation — no frame callback is registered.
  const dir = makeTempDirSync("tn-pump-multi-");
  const bundle = join(dir, "game.js");
  writeFileSync(
    bundle,
    "let n = 0;\n" +
      "const id = setInterval(() => { n += 1; " +
      "if (n === 6) { const end = Date.now() + 400; while (Date.now() < end) {} } " +
      "if (n >= 12) { clearInterval(id); process.exit(0); } }, 50);\n",
  );
  const payload = pumpLine(runHost(bundle, "plain"));
  assert.ok(payload.pumpCount > 2, `expected many pumps, got ${payload.pumpCount}`);
  assert.ok(payload.maxGapMs >= 250, `inter-entry stall lost: ${JSON.stringify(payload)}`);
  assert.ok(
    payload.longGaps.some((gap) => gap >= 250),
    `no retained long gap: ${JSON.stringify(payload)}`,
  );
});

test("real host: shutdown without a present still flushes", () => {
  // A timer-only run never presents, so the first-frame endpoint never fires;
  // the main-loop-exit flush is the bounded endpoint. Missing line fails.
  const dir = makeTempDirSync("tn-pump-noflip-");
  const bundle = join(dir, "game.js");
  writeFileSync(
    bundle,
    "let n = 0;\nconst id = setInterval(() => { n += 1; if (n >= 4) { clearInterval(id); process.exit(0); } }, 50);\n",
  );
  const payload = pumpLine(runHost(bundle, "plain"));
  assert.equal(payload.observed, true);
  assert.ok(payload.firstPumpAtMs >= 0, "process-to-first-pump stamped");
  assert.ok(payload.trailingGapMs >= 0, "trailing interval measured at exit");
});

test("real host: SIGKILL leaves no endpoint line — missing must fail", async () => {
  const log = await runHostKill("setInterval(() => {}, 50);\n");
  assert.ok(log.includes("game_eval_begin"), "the killed run really started");
  assert.ok(
    !log.includes("TN_PUMP_SILENCE:"),
    "a killed run must not produce an observation to misread as good",
  );
  assert.throws(() => pumpLine(log), /TN_PUMP_SILENCE_UNOBSERVED/u);
});

test("real host: mailbox respond() yields a correlated endpoint snapshot", () => {
  // Full option-(a) path on the real host: the game reports through the
  // existing mailbox, the host stamps atMs on the launch clock, counters
  // survive for the later loop-exit flush.
  const { log, res } = runMailboxRespond();
  const endpoints = endpointSnapshots(log);
  assert.ok(endpoints.length >= 1, "respond() must emit TN_PUMP_ENDPOINT");
  const endpoint = endpoints.at(-1);
  assert.equal(endpoint.path, res, "endpoint correlates to the actual response path");
  assert.equal(endpoint.stored, true);
  const { pump } = endpoint;
  assert.equal(pump.observed, true);
  assert.ok(pump.atMs > pump.lastPumpAtMs, "endpoint covers the report, not a stale stamp");
  assert.ok(pump.firstPumpAtMs >= 0, "process-to-first-pump retained through endpoint");
  const diagnostics = (log.match(/TN_PUMP_SILENCE:/g) ?? []).length;
  assert.equal(diagnostics, 1, "requested snapshot must not suppress the diagnostic line");
});

test("real host: >250ms trailing stall through the endpoint is rejected-worthy", () => {
  // 400 ms spin inside the pump just before respond(): the endpoint's
  // trailing interval carries it. The verifier (not this file) rejects;
  // here we prove the number actually arrives on the real host.
  const { log } = runMailboxRespond({ stallBeforeRespondMs: 400 });
  const endpoint = endpointSnapshots(log).at(-1);
  assert.ok(endpoint, "stall run must still produce the endpoint");
  assert.ok(
    endpoint.pump.trailingGapMs >= 250,
    `trailing stall lost: ${JSON.stringify(endpoint.pump)}`,
  );
});
